From 2d4df3405e07922a5c38b7a566853c6b3dc0e7b7 Mon Sep 17 00:00:00 2001
From: Antonio Borneo <antonio.borneo@foss.st.com>
Date: Mon, 4 Mar 2024 12:06:56 +0100
Subject: [PATCH 0774/1695] clk: stm32mp25: program c1msrd during suspend

The register RCC_C1MSRDCR is used by bootrom at D1 Standby exit
but can be reset by HW, e.g. during Standby.

Program RCC_C1MSRDCR register during each suspend, instead of
programming it only once at probe.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git b59bd39a51a30004df61bb32949516313ba46286 clk: stm32mp25: program c1msrd during suspend]
Signed-off-by: Antonio Borneo <antonio.borneo@foss.st.com>
Change-Id: I692e16ccf53224e36dde68e8591b478554d112a7
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/364373
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/374141
Domain-Review: Patrick DELAUNAY <patrick.delaunay@foss.st.com>
Tested-by: Gabriel FERNANDEZ <gabriel.fernandez@foss.st.com>
Reviewed-by: Gabriel FERNANDEZ <gabriel.fernandez@foss.st.com>
Reviewed-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
---
 core/drivers/clk/clk-stm32mp25.c | 10 ++++++----
 1 file changed, 6 insertions(+), 4 deletions(-)

diff --git a/core/drivers/clk/clk-stm32mp25.c b/core/drivers/clk/clk-stm32mp25.c
index 676828c82..94ef2faee 100644
--- a/core/drivers/clk/clk-stm32mp25.c
+++ b/core/drivers/clk/clk-stm32mp25.c
@@ -4279,10 +4279,6 @@ static void clk_stm32_init_oscillators(const void *fdt, int node)
 
 static TEE_Result clk_stm32_apply_rcc_config(struct stm32_clk_platdata *pdata)
 {
-	if (pdata->c1msrd != UINT32_MAX)
-		io_write32(pdata->rcc_base + RCC_C1MSRDCR, pdata->c1msrd &
-			   RCC_C1MSRDCR_C1MSRD_MASK);
-
 	if (pdata->safe_rst)
 		stm32mp25_syscfg_set_safe_reset(true);
 
@@ -4736,8 +4732,14 @@ static TEE_Result stm32_rcc_pm_resume(void)
 
 static TEE_Result stm32_rcc_pm_suspend(void)
 {
+	struct stm32_clk_platdata *pdata = &stm32mp25_clock_pdata;
+
 	clk_save_context();
 
+	/* Set c1msrd for bootrom use. It's reset by HW during standby */
+	io_write32(pdata->rcc_base + RCC_C1MSRDCR, pdata->c1msrd &
+		   RCC_C1MSRDCR_C1MSRD_MASK);
+
 	return TEE_SUCCESS;
 }
 
-- 
2.47.2

